scanpy/preprocessing/_highly_variable_genes.py

```diff
@@ -175,6 +176,196 @@ def _highly_variable_genes_seurat_v3(
         return df
 
 
+def _highly_variable_pearson_residuals(
+    adata: AnnData,
+    layer: Optional[str] = None,
+    n_top_genes: int = 2000,
+    batch_key: Optional[str] = None,
+    theta: float = 100,
+    clip: Union[Literal['auto', 'none'], float] = 'auto',
+    chunksize: int = 100,
+    subset: bool = False,
+    inplace: bool = True,
+) -> Optional[pd.DataFrame]:
+    """\
+    See `highly_variable_genes`.
+
+    Returns
+    -------
+    Depending on `inplace` returns calculated metrics (:class:`~pd.DataFrame`)
+    or updates `.var` with the following fields
+
+    highly_variable
+        boolean indicator of highly-variable genes
+    means
+        means per gene
+    variances
+        variances per gene 
+    residual_variances
+        Pearson residual variance per gene. Averaged in the case of multiple
+        batches.
+    highly_variable_rank
+        Rank of the gene according to residual variance, median rank in the
+        case of multiple batches
+    highly_variable_nbatches : int
+        If batch_key is given, this denotes in how many batches genes are
+        detected as HVG
+    highly_variable_intersection : bool
+        If batch_key is given, this denotes the genes that are highly variable
+        in all batches
+    """
+
+    view_to_actual(adata)
+    X = _get_obs_rep(adata, layer=layer)
+    computed_on = layer if layer else 'adata.X'
+
+    # Check for raw counts
+    if check_nonnegative_integers(X) is False:
+        raise ValueError(
+            "`pp.highly_variable_genes` with `flavor='pearson_residuals'`"
+            "expects raw count data."
+        )
+
+    if batch_key is None:
```

Comment:
I see this is the same as seurat_V3 but don't understand the need for transforming it in `pd.Categorical` if, in the case `batch_key` is passed, becomes `np.ndarray`

Reply:
True. I removed it now..

Maybe a residual from older days? In the main function, `adata.obs.['batch']` seems to expect `pd.Categorical` though:

https://github.com/theislab/scanpy/blob/af8f323ebca87bc98d51e556742acf3c2cdf56e9/scanpy/preprocessing/_highly_variable_genes.py#L446